### PR TITLE
fix: early return for module file error in AST parsing

### PR DESCRIPTION
**Pull request by ghost**
Thank you for making Rust GCC better!

If your PR fixes an issue, you can add "Fixes #issue_number" into this
PR description and the git commit message. This way the issue will be
automatically closed when your PR is merged. If your change addresses
an issue but does not fully fix it please mark it as "Addresses #issue_number"
in the git commit message.

Here is a checklist to help you with your PR.

- \[x] GCC development requires copyright assignment or the Developer's Certificate of Origin sign-off, see https://gcc.gnu.org/contribute.html or https://gcc.gnu.org/dco.html
- \[x] Read contributing guidlines
- \[x] `make check-rust` passes locally
- \[x] Run `clang-format`
- \[x] Added any relevant test cases to `gcc/testsuite/rust/`

Note that you can skip the above if you are just opening a WIP PR in
order to get feedback.
---

Converts an assert into an early return during AST parsing.

Resolves: https://github.com/Rust-GCC/gccrs/issues/4145

### CODE REVIEW

File: gcc/testsuite/rust/compile/issue-4145.rs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+struct S {
```
Comment (ghost) [inline gcc/testsuite/rust/compile/issue-4145.rs:3]: Not sure on the exact decorators needed here, or if this test should go into the `xfail` directory. The test is expected to fail compilation, just with an error message instead of an ICE.

File: ChangeLog [hunk 2]
```diff
@@ -1,3 +1,8 @@
+2025-09-12  Elle Rhumsaa  <elle@weathered-steel.dev>
```
Comment (powerboat9) [inline ChangeLog]: You shouldn't be updating the ChangeLog file -- there's a script which automatically does it on the GCC master branch

File: gcc/testsuite/rust/compile/issue-4145.rs (hunk 1 above)
Comment (powerboat9) [reply gcc/testsuite/rust/compile/issue-4145.rs:3]: `grep` for `dg-error` in the testsuite directory -- that should give you some nice examples for expecting error messages

File: gcc/rust/ast/rust-ast.cc [hunk 3]
```diff
@@ -3383,7 +3383,13 @@ void
 Module::process_file_path ()
 {
   rust_assert (kind == Module::ModuleKind::UNLOADED);
-  rust_assert (module_file.empty ());
+
+  if (module_file.empty ())
+    {
+      rust_error_at (locus, "error handling module file for %s",
```
Comment (CohenArthur) [inline gcc/rust/ast/rust-ast.cc]:
```suggestion
      rust_error_at (locus, "error handling module file for %qs",
```

This will output a quoted string which looks better and allows the user to better see which path is wrong

File: gcc/testsuite/rust/compile/issue-4145.rs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+// { dg-excess-errors "warnings" }
+
+struct S {
+    field: [u8; {
+        #[path = "outer/inner.rs"]
+        // { dg-error "error handling module file for inner" "#4145" { xfail *-*-* } .-1 }
```
Comment (CohenArthur) [inline gcc/testsuite/rust/compile/issue-4145.rs]:
```suggestion
        // { dg-error "error handling module file for .inner." "#4145" { xfail *-*-* } .-1 }
```

So if you apply the change I suggested above you also need to do this, as the quotes are different depending on your system - e.g. it can be output as `` `inner` `` or `‘inner’`

File: ChangeLog (hunk 2 above)
Comment (P-E-P) [reply ChangeLog]: @0xllx0 You still need the changelog section within your commit message, the changelog file is generated using the content of your commit message.

Comment (P-E-P) [reply ChangeLog]: A script can generate the changelog structure for you. You can install it for this repository using the script `$TOP_LEVEL/contrib/gcc-git-customization.sh`, it will ask you a bunch of things (you can leave all default values) and then it will install a new git command.

You may then use `git gcc-commit-mklog -s` instead of `git commit -s` when making a commit.

Comment (ghost) [reply ChangeLog]: Thanks, I was looking for documentation on which script to run, and how to do it. Are these steps included in any documentation? If not, I'd be happy to write something up.

Comment (P-E-P) [reply ChangeLog]: It is briefly mentioned [here](https://github.com/Rust-GCC/gccrs/wiki/First-contributions#changelog).

**Comment by philberty** [comment]
You need to fixup the regressions here those tests probably emit a new error now and they need updated accordingly

**Comment by ghost** [comment]
> You need to fixup the regressions here those tests probably emit a new error now and they need updated accordingly

I rebased on the latest master (`32ce955`), and didn't get a failing test for the `gcc/testsuite/rust/compile/issue-4145.rs` test. Is there something specific where you are noticing regressions? Can you please provide reproduction steps so I can work on resolving the issue?

File: gcc/testsuite/rust/compile/issue-4145.rs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+// { dg-excess-errors "warnings" }
+
+struct S {
+    field: [u8; {
+        #[path = "outer/inner.rs"]
+        // { dg-error "error handling module file for .inner." "#4145" { xfail *-*-* } .-1 }
```
Comment (powerboat9) [inline gcc/testsuite/rust/compile/issue-4145.rs]: I think `.-1` should be `.1`, if you want to catch an error on the next line instead of the previous line

Comment (ghost) [reply gcc/testsuite/rust/compile/issue-4145.rs]: Thanks, I changed it to `.+1`, and changed `xfail` to `target` since the compiler successfully returns an error (opposed to an ICE). Please let me know if that logic makes sense to you.

**Comment by P-E-P** [comment]
> I rebased on the latest master (`32ce955`), and didn't get a failing test for the `gcc/testsuite/rust/compile/issue-4145.rs` test. Is there something specific where you are noticing regressions? Can you please provide reproduction steps so I can work on resolving the issue?

I can't remember the state of that PR when you first submitted it. This message seems to be out of date (there are multiple tests that are not passing anymore). Do you struggle with it ? Do you need some help ?

You probably need more constraints over your condition. Try executing the test using GDB to understand why it emits an error. Note that the tests may also have a slight chance of being wrong, even though I doubt it.

**Comment by ghost** [comment]
> Do you struggle with it ? Do you need some help ?

I don't know enough at this point about the internals of `gccrs` to determine exactly why the early return is breaking the other tests. The test for `issue-4145.rs` passes, but obviously the other tests break.

Not sure what the best way forward is.

> You probably need more constraints over your condition.

Possibly, but I currently lack the appropriate context for what the test condition should be that captures the motivating edge-case, and doesn't break the rest of the test suite. It will take me some time to get more familiar with the surrounding code.

> Try executing the test using GDB to understand why it emits an error. Note that the tests may also have a slight chance of being wrong, even though I doubt it.

Thanks, that will probably help trace the relevant code paths.

**Comment by ghost** [comment]
> You probably need more constraints over your condition.

Thanks for mentioning this, because it made me review the condition for the early return. Somehow, I inverted the boolean condition at some point. It's fixed now.
